doc/source/whatsnew/v0.21.0.txt

```diff
@@ -506,6 +506,7 @@ Indexing
 - Bug in ``CategoricalIndex`` reindexing in which specified indices containing duplicates were not being respected (:issue:`17323`)
 - Bug in intersection of ``RangeIndex`` with negative step (:issue:`17296`)
 - Bug in ``IntervalIndex`` where performing a scalar lookup fails for included right endpoints of non-overlapping monotonic decreasing indexes (:issue:`16417`, :issue:`17271`)
+- Bug in adding ``DatetimeIndex`` with a ``TimedeltaIndex`` or a numpy array with ``dtype="timedelta64"`` (:issue:`17558`)
```

Comment:
this is not correct; the bug is *only* with a numpy array of ``dtype=timedelta64[ns]`` (timedelta64 is not a valid dtype for pandas)

Reply:
The "or" was a bit confusing, but there are two separate bugs:
* adding tz-aware ``DatetimeIndex`` with ``TimedeltaIndex``
* adding tz-aware ``DatetimeIndex`` with numpy array which is timedelta64-dtype-like

Reply:
how so? where does DTI + TDI tz-aware not work? your issues is w.r.t. np.array of timedeltas

Reply:
I mentioned it briefly in the issue ticket, although I neglected to post the full stack trace:

```python
In [1]: import numpy as np, pandas as pd

In [2]: dti = pd.DatetimeIndex([pd.Timestamp("2017/01/01")]).tz_localize('US/Eastern')

In [3]: td_np = np.array([np.timedelta64(1, 'ns')]) 

In [4]: (dti + pd.TimedeltaIndex(td_np))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-20f28143c2bb> in <module>()
----> 1 (dti + pd.TimedeltaIndex(td_np))

/spare/local/azhu/venv-el6-20170811-stonelib/lib/python2.7/site-packages/pandas/tseries/base.pyc in __add__(self, other)
    617             from pandas.tseries.offsets import DateOffset
    618             if isinstance(other, TimedeltaIndex):
--> 619                 return self._add_delta(other)
    620             elif isinstance(self, TimedeltaIndex) and isinstance(other, Index):
    621                 if hasattr(other, '_add_delta'):

/spare/local/azhu/venv-el6-20170811-stonelib/lib/python2.7/site-packages/pandas/tseries/index.pyc in _add_delta(self, delta)
    787             new_values = self._add_delta_td(delta)
    788         elif isinstance(delta, TimedeltaIndex):
--> 789             new_values = self._add_delta_tdi(delta)
    790             # update name when delta is Index
    791             name = com._maybe_match_name(self, delta)

/spare/local/azhu/venv-el6-20170811-stonelib/lib/python2.7/site-packages/pandas/tseries/base.pyc in _add_delta_tdi(self, other)
    704             mask = (self._isnan) | (other._isnan)
    705             new_values[mask] = tslib.iNaT
--> 706         return new_values.view(self.dtype)
    707 
    708     def isin(self, values):

TypeError: data type not understood
```